src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -78,17 +127,25 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
 
     fn variant_field(
         &self,
-        path_str: &str,
+        path_str: &'path str,
         current_item: &Option<String>,
         module_id: DefId,
-    ) -> Result<(Res, Option<String>), ErrorKind> {
+        extra_fragment: &Option<String>,
+    ) -> Result<(Res, Option<String>), ErrorKind<'path>> {
         let cx = self.cx;
 
+        debug!("looking for enum variant {}", path_str);
         let mut split = path_str.rsplitn(3, "::");
-        let variant_field_name =
-            split.next().map(|f| Symbol::intern(f)).ok_or(ErrorKind::ResolutionFailure)?;
+        let variant_field_name = split
+            .next()
+            .map(|f| Symbol::intern(f))
+            .expect("fold_item should ensure link is non-empty");
```

Comment:
Based on the description, resolving...

Reply:
> The previous behavior might have been masking incorrect rustc behavior, but it would now be an ICE

Right, and that's intentional because it's a bug. There's no reasonable error we could give here other than 'rustdoc has buggy logic', which I think is better served by an ICE.